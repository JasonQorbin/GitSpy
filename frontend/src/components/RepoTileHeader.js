import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faArrowUpRightFromSquare} from "@fortawesome/free-solid-svg-icons";

/**
 * Function component that displays the top half of the RepoTile header
 *
 * @param props Must contain the repo object from github.
 * @returns {JSX.Element}
 * @constructor
 */
function RepoTileHeader(props) {
    const dateOptions = {
        year: "numeric",
        month: "short",
        day: "numeric"
    }
    return (
        <div className="repo-tile-row-wrapper">
            <div className="left-tile-pane">
                <h3><a className="external-link" href={props.repo.html_url}>{props.repo.name}<FontAwesomeIcon
                    icon={faArrowUpRightFromSquare}/></a></h3>
                <p><em>{props.repo.description}</em></p>
            </div>
            <div className="right-tile-pane">
                <p>Languages: <strong>{props.repo.language}</strong></p>
                <p><em>Created: {props.createdAt.toLocaleString(undefined, dateOptions)}</em></p>
                <p><em>Last commit: {props.updatedAt.toLocaleString(undefined, dateOptions)}</em></p>
            </div>
        </div>
    )
}

export default RepoTileHeader;