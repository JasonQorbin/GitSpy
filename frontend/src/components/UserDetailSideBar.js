import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faArrowUpRightFromSquare, faEnvelope, faMapPin,  } from "@fortawesome/free-solid-svg-icons";
import { faXTwitter } from "@fortawesome/free-brands-svg-icons";

/**
 * Functional component that renders the contents of the aside element of the User Detail page.
 *
 * Not every GitHub user fills in all their information or may have chosen to not make certain fields
 * like their email publicly visible. This component checks if the information for a field is available before
 * rendering it section for it.
 *
 * @param props Should contain all the fields to be displayed. If one of the fields is not available the value should be null.
 * @returns {JSX.Element}
 * @constructor
 */
function UserDetailSideBar(props) {
    const elements = [];

    if (props.profilePic === null) {
        elements.push(<div key="profile-pic" id="profile-pic"></div>);
    } else {
        elements.push(<img key="profile-pic" id="profile-pic" src={props.profilePic}/>);
    }

    elements.push(<h1 key="fullName">{props.fullName}</h1>);
    elements.push(<h2 key="username">
        <a className="external-link" href={props.userURL}>{props.username}<FontAwesomeIcon icon={faArrowUpRightFromSquare} /></a>
    </h2>);
    elements.push(<span key="divider" className="divider"></span>);
    if (props.location !== null) {
        elements.push(<p key="location"><FontAwesomeIcon icon={faMapPin}/>{props.location}</p>)
    }

    if (props.email !== null) {
        elements.push(<p key="email"><FontAwesomeIcon icon={faEnvelope}/>{props.email}</p>);
    }

    if (props.twitter !== null) {
        elements.push(<p key="twitter"><FontAwesomeIcon icon={faXTwitter} />{props.twitter}</p>);
    }

    return (
        <div className="user-detail-section">
            {elements}
        </div>
    );
}

export default UserDetailSideBar;