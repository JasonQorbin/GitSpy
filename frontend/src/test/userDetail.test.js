import React from 'react'
import renderer from 'react-test-renderer';
import UserDetailSideBar from "../components/UserDetailSideBar";
import RepoTileHeader from "../components/RepoTileHeader";

test('Full User detail side bar', () => {
    const tree = renderer
        .create(
            <UserDetailSideBar
                profilePic="testImageURL"
                fullName="testFullName"
                username="testUsername"
                location="testLocation"
                email="testEmail"
                twitter="testEmail"
                userURL="testUserURL"/>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

test('User detail side bar without email, location or twitter', () => {
    const tree = renderer
        .create(
            <UserDetailSideBar
                profilePic="testImageURL"
                fullName="testFullName"
                username="testUsername"
                userURL="testUserURL"/>)
        .toJSON();
    expect(tree).toMatchSnapshot();
})

test("Repo tile header", () => {
    const testData = [
        {
            updated_at : "2023-12-04T20:43:40Z",
            created_at : "2021-12-04T20:43:40Z",
            id: 11223344,
            html_url: "testHTMLurl",
            name: "testName",
            description: "Some test description",
            language: "Haskell",
        }
    ];

    const updatedAt = new Date(testData.updated_at);
    const createdAt = new Date(testData.created_at);
    const tree = renderer
        .create(<RepoTileHeader repo={testData} updatedAt={updatedAt} createdAt={createdAt}/>)
        .toJSON();
    expect(tree).toMatchSnapshot();
})
