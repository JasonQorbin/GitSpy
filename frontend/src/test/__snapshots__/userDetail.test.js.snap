// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Full User detail side bar 1`] = `
<div
  className="user-detail-section"
>
  <img
    id="profile-pic"
    src="testImageURL"
  />
  <h1>
    testFullName
  </h1>
  <h2>
    <a
      className="external-link"
      href="testUserURL"
    >
      testUsername
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-arrow-up-right-from-square "
        data-icon="arrow-up-right-from-square"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32h82.7L201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3V192c0 17.7 14.3 32 32 32s32-14.3 32-32V32c0-17.7-14.3-32-32-32H320zM80 32C35.8 32 0 67.8 0 112V432c0 44.2 35.8 80 80 80H400c44.2 0 80-35.8 80-80V320c0-17.7-14.3-32-32-32s-32 14.3-32 32V432c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V112c0-8.8 7.2-16 16-16H192c17.7 0 32-14.3 32-32s-14.3-32-32-32H80z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </a>
  </h2>
  <span
    className="divider"
  />
  <p>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-map-pin "
      data-icon="map-pin"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 144a144 144 0 1 1 288 0A144 144 0 1 1 16 144zM160 80c8.8 0 16-7.2 16-16s-7.2-16-16-16c-53 0-96 43-96 96c0 8.8 7.2 16 16 16s16-7.2 16-16c0-35.3 28.7-64 64-64zM128 480V317.1c10.4 1.9 21.1 2.9 32 2.9s21.6-1 32-2.9V480c0 17.7-14.3 32-32 32s-32-14.3-32-32z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    testLocation
  </p>
  <p>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-envelope "
      data-icon="envelope"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M48 64C21.5 64 0 85.5 0 112c0 15.1 7.1 29.3 19.2 38.4L236.8 313.6c11.4 8.5 27 8.5 38.4 0L492.8 150.4c12.1-9.1 19.2-23.3 19.2-38.4c0-26.5-21.5-48-48-48H48zM0 176V384c0 35.3 28.7 64 64 64H448c35.3 0 64-28.7 64-64V176L294.4 339.2c-22.8 17.1-54 17.1-76.8 0L0 176z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    testEmail
  </p>
  <p>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-x-twitter "
      data-icon="x-twitter"
      data-prefix="fab"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M389.2 48h70.6L305.6 224.2 487 464H345L233.7 318.6 106.5 464H35.8L200.7 275.5 26.8 48H172.4L272.9 180.9 389.2 48zM364.4 421.8h39.1L151.1 88h-42L364.4 421.8z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    testEmail
  </p>
</div>
`;

exports[`Repo tile header 1`] = `
<div
  className="repo-tile-row-wrapper"
>
  <div
    className="left-tile-pane"
  >
    <h3>
      <a
        className="external-link"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-arrow-up-right-from-square "
          data-icon="arrow-up-right-from-square"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32h82.7L201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3V192c0 17.7 14.3 32 32 32s32-14.3 32-32V32c0-17.7-14.3-32-32-32H320zM80 32C35.8 32 0 67.8 0 112V432c0 44.2 35.8 80 80 80H400c44.2 0 80-35.8 80-80V320c0-17.7-14.3-32-32-32s-32 14.3-32 32V432c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V112c0-8.8 7.2-16 16-16H192c17.7 0 32-14.3 32-32s-14.3-32-32-32H80z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </h3>
    <p>
      <em />
    </p>
  </div>
  <div
    className="right-tile-pane"
  >
    <p>
      Languages: 
      <strong />
    </p>
    <p>
      <em>
        Created: 
        Invalid Date
      </em>
    </p>
    <p>
      <em>
        Last commit: 
        Invalid Date
      </em>
    </p>
  </div>
</div>
`;

exports[`Repo tiles 1`] = `
<div>
  <div
    className="repo-tile"
  >
    <div
      className="left-tile-pane"
    >
      <h3>
        <a
          className="external-link"
          href="testHTMLurl"
        >
          testName
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-arrow-up-right-from-square "
            data-icon="arrow-up-right-from-square"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32h82.7L201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3V192c0 17.7 14.3 32 32 32s32-14.3 32-32V32c0-17.7-14.3-32-32-32H320zM80 32C35.8 32 0 67.8 0 112V432c0 44.2 35.8 80 80 80H400c44.2 0 80-35.8 80-80V320c0-17.7-14.3-32-32-32s-32 14.3-32 32V432c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V112c0-8.8 7.2-16 16-16H192c17.7 0 32-14.3 32-32s-14.3-32-32-32H80z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </a>
      </h3>
      <p>
        <em>
          Some test description
        </em>
      </p>
    </div>
    <div
      className="right-tile-pane"
    >
      <p>
        Languages: 
        <strong>
          Haskell
        </strong>
      </p>
      <p>
        <em>
          Last commit: 
          4 Dec 2023
        </em>
      </p>
    </div>
  </div>
  <div
    className="repo-tile"
  >
    <div
      className="left-tile-pane"
    >
      <h3>
        <a
          className="external-link"
          href="testHTMLurl2"
        >
          testName2
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-arrow-up-right-from-square "
            data-icon="arrow-up-right-from-square"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32h82.7L201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3V192c0 17.7 14.3 32 32 32s32-14.3 32-32V32c0-17.7-14.3-32-32-32H320zM80 32C35.8 32 0 67.8 0 112V432c0 44.2 35.8 80 80 80H400c44.2 0 80-35.8 80-80V320c0-17.7-14.3-32-32-32s-32 14.3-32 32V432c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V112c0-8.8 7.2-16 16-16H192c17.7 0 32-14.3 32-32s-14.3-32-32-32H80z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </a>
      </h3>
      <p>
        <em>
          Some other test description
        </em>
      </p>
    </div>
    <div
      className="right-tile-pane"
    >
      <p>
        Languages: 
        <strong>
          Pascal
        </strong>
      </p>
      <p>
        <em>
          Last commit: 
          24 Sept 2023
        </em>
      </p>
    </div>
  </div>
  <div
    className="repo-tile"
  >
    <div
      className="left-tile-pane"
    >
      <h3>
        <a
          className="external-link"
          href="testHTMLurl3"
        >
          testName3
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-arrow-up-right-from-square "
            data-icon="arrow-up-right-from-square"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32h82.7L201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3V192c0 17.7 14.3 32 32 32s32-14.3 32-32V32c0-17.7-14.3-32-32-32H320zM80 32C35.8 32 0 67.8 0 112V432c0 44.2 35.8 80 80 80H400c44.2 0 80-35.8 80-80V320c0-17.7-14.3-32-32-32s-32 14.3-32 32V432c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V112c0-8.8 7.2-16 16-16H192c17.7 0 32-14.3 32-32s-14.3-32-32-32H80z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </a>
      </h3>
      <p>
        <em>
          Yet another test description
        </em>
      </p>
    </div>
    <div
      className="right-tile-pane"
    >
      <p>
        Languages: 
        <strong>
          Ruby
        </strong>
      </p>
      <p>
        <em>
          Last commit: 
          4 Mar 2021
        </em>
      </p>
    </div>
  </div>
</div>
`;

exports[`User detail side bar without email, location or twitter 1`] = `
<div
  className="user-detail-section"
>
  <img
    id="profile-pic"
    src="testImageURL"
  />
  <h1>
    testFullName
  </h1>
  <h2>
    <a
      className="external-link"
      href="testUserURL"
    >
      testUsername
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-arrow-up-right-from-square "
        data-icon="arrow-up-right-from-square"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M320 0c-17.7 0-32 14.3-32 32s14.3 32 32 32h82.7L201.4 265.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L448 109.3V192c0 17.7 14.3 32 32 32s32-14.3 32-32V32c0-17.7-14.3-32-32-32H320zM80 32C35.8 32 0 67.8 0 112V432c0 44.2 35.8 80 80 80H400c44.2 0 80-35.8 80-80V320c0-17.7-14.3-32-32-32s-32 14.3-32 32V432c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V112c0-8.8 7.2-16 16-16H192c17.7 0 32-14.3 32-32s-14.3-32-32-32H80z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </a>
  </h2>
  <span
    className="divider"
  />
  <p>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-map-pin "
      data-icon="map-pin"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 144a144 144 0 1 1 288 0A144 144 0 1 1 16 144zM160 80c8.8 0 16-7.2 16-16s-7.2-16-16-16c-53 0-96 43-96 96c0 8.8 7.2 16 16 16s16-7.2 16-16c0-35.3 28.7-64 64-64zM128 480V317.1c10.4 1.9 21.1 2.9 32 2.9s21.6-1 32-2.9V480c0 17.7-14.3 32-32 32s-32-14.3-32-32z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </p>
  <p>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-envelope "
      data-icon="envelope"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M48 64C21.5 64 0 85.5 0 112c0 15.1 7.1 29.3 19.2 38.4L236.8 313.6c11.4 8.5 27 8.5 38.4 0L492.8 150.4c12.1-9.1 19.2-23.3 19.2-38.4c0-26.5-21.5-48-48-48H48zM0 176V384c0 35.3 28.7 64 64 64H448c35.3 0 64-28.7 64-64V176L294.4 339.2c-22.8 17.1-54 17.1-76.8 0L0 176z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </p>
  <p>
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-x-twitter "
      data-icon="x-twitter"
      data-prefix="fab"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M389.2 48h70.6L305.6 224.2 487 464H345L233.7 318.6 106.5 464H35.8L200.7 275.5 26.8 48H172.4L272.9 180.9 389.2 48zM364.4 421.8h39.1L151.1 88h-42L364.4 421.8z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </p>
</div>
`;
